server/sonar-process-monitor/src/main/java/org/sonar/process/monitor/Monitor.java

```diff
@@ -98,54 +127,246 @@ private void monitor(ProcessRef processRef) {
+  }
+
+  private void stopAsync(State stoppingState) {
+    if (lifecycle.tryToMoveTo(stoppingState)) {
       terminator.start();
     }
   }
 
+  public void restartAsync() {
+    if (lifecycle.tryToMoveTo(State.RESTARTING)) {
+      restartor = new RestartorThread();
+      restartor.start();
+    }
+  }
+
+  /**
+   * Runs every time a restart request is detected.
+   */
+  private class RestartorThread extends Thread {
+
+    private RestartorThread() {
+      super("Restartor " + (restartorInstanceCounter++));
+    }
+
+    @Override
+    public void run() {
+      stopProcesses();
+      startProcesses();
+    }
+  }
+
+  /**
+   * Runs only once
+   */
+  private class TerminatorThread extends Thread {
+
+    private TerminatorThread() {
+      super("Terminator");
+    }
+
+    @Override
+    public void run() {
+      stopProcesses();
+    }
+  }
+
+  /**
+   * Watches for any child process requesting a restart of all children processes.
+   * It runs once and as long as {@link #lifecycle} hasn't reached {@link Lifecycle.State#STOPPED} and holds its checks
+   * when {@link #lifecycle} is not in state {@link Lifecycle.State#STARTED} to avoid taking the same request into account
+   * twice.
+   */
+  public class RestartRequestWatcherThread extends Thread {
+    public RestartRequestWatcherThread() {
+      super("Restart watcher");
+    }
+
+    @Override
+    public void run() {
+      while (lifecycle.getState() != Lifecycle.State.STOPPED) {
+        if (lifecycle.getState() == Lifecycle.State.STARTED && didAnyProcessRequestRestart()) {
+          restartAsync();
+        }
+        try {
+          Thread.sleep(WATCH_DELAY_MS);
+        } catch (InterruptedException ignored) {
+          // keep watching
+        }
+      }
+    }
+
+    private boolean didAnyProcessRequestRestart() {
+      for (WatcherThread watcherThread : watcherThreads) {
+        ProcessRef processRef = watcherThread.getProcessRef();
+        if (processRef.getCommands().askedForRestart()) {
+          LOG.info("Process [{}] requested restart", processRef.getKey());
+          return true;
+        }
+      }
+      return false;
+    }
+
+  }
+
+  public class HardStopWatcherThread extends Thread {
+    private final ProcessCommands processCommands;
+
+    public HardStopWatcherThread(ProcessCommands processCommands) {
+      super("Hard stop watcher");
+      this.processCommands = processCommands;
+    }
+
+    @Override
+    public void run() {
+      while (lifecycle.getState() != Lifecycle.State.STOPPED) {
+        if (processCommands.askedForStop()) {
+          trace("Stopping process");
+          Monitor.this.stop();
+        } else {
+          try {
+            Thread.sleep(WATCH_DELAY_MS);
+          } catch (InterruptedException ignored) {
+            // keep watching
+          }
+        }
+      }
+    }
+
+  }
+
+  private void stopProcesses() {
+    ArrayList<WatcherThread> watcherThreads = new ArrayList<>(this.watcherThreads);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Rename "watcherThreads" which hides the field declared at line 52. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AHiddenFieldCheck)